container.go

```diff
@@ -3,79 +3,67 @@
 package container
 
 import (
+	"context"
 	"errors"
 	"fmt"
 	"reflect"
 	"unsafe"
 )
 
-// binding holds a resolver and a concrete (if already resolved).
-// It is the break for the Container wall!
-type binding struct {
-	resolver    interface{} // resolver is the function that is responsible for making the concrete.
-	concrete    interface{} // concrete is the stored instance for singleton bindings.
-	isSingleton bool        // isSingleton is true if the binding is a singleton.
+// Container holds the bindings and provides methods to interact with them.
+// It is the entry point in the package.
+type Container struct {
+	parent   *Container
+	bindings map[reflect.Type]map[string]*binding
 }
 
-// make resolves the binding if needed and returns the resolved concrete.
-func (b *binding) make(c Container) (interface{}, error) {
-	if b.concrete != nil {
-		return b.concrete, nil
-	}
-
-	retVal, err := c.invoke(b.resolver)
-	if b.isSingleton {
-		b.concrete = retVal
+// New creates a new instance of the Container.
+func New() *Container {
+	return &Container{
+		bindings: make(map[reflect.Type]map[string]*binding),
 	}
-
-	return retVal, err
 }
 
-// Container holds the bindings and provides methods to interact with them.
-// It is the entry point in the package.
-type Container map[reflect.Type]map[string]*binding
+func (c *Container) NewScope() (*Container, error) {
+	childContainer := New()
+	childContainer.parent = c
 
-// New creates a new concrete of the Container.
-func New() Container {
-	return make(Container)
+	for _, outerBinding := range c.bindings {
+		for name, binding := range outerBinding {
+			if binding.lifetime == Scoped {
+				if err := childContainer.bind(binding.resolver, name, Singleton); err != nil {
```

Comment:
Should this be `Scoped` or do we want to change the lifetime to `Singleton` when binding in this child container?  It seems a little strange to me that it would become a singleton?

Reply:
A scoped registration act like a singleton but only within the lifetime of that scope.  So basically when a new scope is created we transition that binding to a singleton binding only within that scope.  The binding itself holds the reference to the initialized concrete instance that is returned on subsequent requests for that type from within the current scoped container.

Reply:
In my mind, the a Scoped lifetime was something where like a singleton we'd only call the resolver once and cache the value, but you could call `NewScope` and get a new container where the bindings for scoped instances were "reset" and so the next resolve would run the resolver function again.

In my mind, this should happen even if you call `NewScope` on a container which itself was returned from `NewScope`.

I think a test like this should pass:

```go
func TestScopedNewScope(t *testing.T) {
	c := New()
	called := 0

	err := c.RegisterScoped(func() *shape { called++; return &shape{} })
	require.NoError(t, err)
	require.Equal(t, 0, called)

	// Resolve the same type twice, since it's scoped, the resolver should only one once, the second time we just return the cached instance.
	var resolved *shape
	err = c.Resolve(&resolved)
	require.NoError(t, err)
	require.Equal(t, 1, called)
	var resolved2 *shape
	err = c.Resolve(&resolved2)
	require.NoError(t, err)
	require.Equal(t, 1, called)
	require.Same(t, resolved, resolved2)

	// Create a new scope, and then resolve in that scope, the resolver should be called again, since we are in a new scope.
	sub, err := c.NewScope()
	require.NoError(t, err)
	var resolved3 *shape
	err = sub.Resolve(&resolved3)
	require.NoError(t, err)
	require.Equal(t, 2, called)
	var resolved4 *shape
	err = sub.Resolve(&resolved4)
	require.NoError(t, err)
	require.Equal(t, 2, called)
	require.Same(t, resolved3, resolved4)

	// Now, create a scope from this container we got from NewScope on the previous container and run the resolvers again.
	sub2, err := sub.NewScope()
	require.NoError(t, err)
	var resolved5 *shape
	err = sub2.Resolve(&resolved5)
	require.NoError(t, err)
	require.Equal(t, 3, called)
	var resolved6 *shape
	err = sub2.Resolve(&resolved6)
	require.NoError(t, err)
	require.Equal(t, 3, called)
	require.Same(t, resolved5, resolved6)
}

type shape struct{}
```

Today it does not (I am worried that the shape of this code here leads to cases where the resolver would not run the 3rd them), FWIW - when I pulled down the `azd` branch and ran this test, it looks like the resolver was running again for the resolve into `resolved2`.

I think I might have a misunderstanding of what `Scoped` should do - do you think this test should pass or is my mental model here wrong?

Reply:
You are correct.  The following statements should be true.

- The root container is inherently a scope.
- Child containers (scopes) should recreate scoped bindings at each scope level
- Resolutions of a scoped binding within the same scope should resolve to the same instance.

Reply:
I've made some updates and added this test case to the test list.